### PR TITLE
New user feedback survey

### PR DESCRIPTION
**Pull request by digitarald**
As we are working to improve first-time setup experience of VS Code, we need to make it easier for new users to tell us about what's working and what is not working for them.

Preliminary heuristics:

- Wait at least 1hr after VS Code's initial startup before showing the survey, but only show it within the first 24hrs of usage.
- Wait for at least 5min after startup to show the survey.
- When a user asks to *remind later*, wait for at least 4hrs. This ensures that we still focus on the first-time experience.
- Drop *Never show again* button in favor of disabling the survey when the user closes the prompt.
- Probability/eligibility is controlled via sampling in experimentation.

The timings above are a starting point and will evolve as we learn.

### CODE REVIEW

**Comment by bpasero** [comment]
I think this is modeled after our existing NPS survey code which I think @joaomoreno [added](https://github.com/microsoft/vscode/commits/f76e090063fc2dc4c298c4b84b56e8a2bc82df68/src/vs/workbench/electron-browser/nps.contribution.ts) so he can maybe also review.

Overall I feel the `TODO` should be addressed and maybe the `setTimeout` put into a `RunOnceScheduler` that gets disposed when the workbench contribution gets disposed.

File: src/vs/workbench/contrib/surveys/browser/ces.contribution.ts [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+			const message = await this.tasExperimentService?.getTreatment<string>('CESSurveyMessage') ?? nls.localize('cesSurveyQuestion', 'Got a moment to help the VS Code team? Please tell us about your experience with VS Code so far.');
+			const button = await this.tasExperimentService?.getTreatment<string>('CESSurveyButton') ?? nls.localize('giveFeedback', "Give Feedback");
+
+			this.notificationService.prompt(
+				Severity.Info,
+				message,
+				[{
+					label: button,
+					run: () => {
+						sendTelemetry('neverShowAgain');
```
Comment (joaomoreno) [inline src/vs/workbench/contrib/surveys/browser/ces.contribution.ts:69]:
```suggestion
						sendTelemetry('accept');
```

Right?

File: src/vs/workbench/contrib/surveys/browser/ces.contribution.ts [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+					"userReaction" : { "classification": "SystemMetaData", "purpose": "FeatureInsight" }
+				}
+				*/
+				this.telemetryService.publicLog('cesSurvey:popup', { userReaction });
+			};
+
+			const message = await this.tasExperimentService?.getTreatment<string>('CESSurveyMessage') ?? nls.localize('cesSurveyQuestion', 'Got a moment to help the VS Code team? Please tell us about your experience with VS Code so far.');
+			const button = await this.tasExperimentService?.getTreatment<string>('CESSurveyButton') ?? nls.localize('giveFeedback', "Give Feedback");
+
+			this.notificationService.prompt(
```
Comment (joaomoreno) [inline src/vs/workbench/contrib/surveys/browser/ces.contribution.ts:63]: The result of `prompt()` should be stored in a variable, such that we can `await` for the prompt to be closed and make sure the `RunOnceScheduler` does its job of just running this one prompt at a time.

File: src/vs/workbench/contrib/surveys/browser/ces.contribution.ts [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+					run: () => {
+						sendTelemetry('neverShowAgain');
+						this._skipSurvey();
+					}
+				}],
+				{ sticky: true }
+			);
+		}, WAIT_TIME_TO_SHOW_SURVEY));
+
+		this._prepareSurvey();
```
Comment (joaomoreno) [inline src/vs/workbench/contrib/surveys/browser/ces.contribution.ts:93]: This should be awaited.

**Review by joaomoreno — CHANGES_REQUESTED** [review]

File: src/vs/workbench/contrib/surveys/browser/ces.contribution.ts (hunk 3 above)
Comment (digitarald) [reply src/vs/workbench/contrib/surveys/browser/ces.contribution.ts:93]: How would I `await` in a `constructor`, @joaomoreno ? This is why I moved the `await` based code chunk into its own function.

Comment (joaomoreno) [reply src/vs/workbench/contrib/surveys/browser/ces.contribution.ts:93]: Oh I misread: I thought it was inside the scheduler. My bad.

**Comment by joaomoreno** [comment]
I see. The `RunOnceScheduler` might actually be limiting you in functionality. What you wanna use is the `ThrottledDelayer`: it provides debouncing **and** awaits for the async function to complete. I've pushed a commit on top which refactors the code to use it, which also improves readability. Give it a test.

**Comment by digitarald** [comment]
@joaomoreno did take a shot based on your commit (and cleaned up some of the naming).

**Review by joaomoreno — APPROVED** [review]

**Comment by joaomoreno** [comment]
Pushed another smaller change: there was a bad typing on that optional service import, it could potentially be `undefined` since it's `@optional`.

**Comment by TylerLeonhardt** [comment]
We talked offline, the risk seems small from what I gather as this will be off by default on stable.

**Pull request merged**